Source/WebCore/platform/graphics/gstreamer/eme/CDMOpenCDM.cpp

```diff
@@ -214,71 +354,109 @@ static RefPtr<SharedBuffer> parseResponseMessage(const String& message)
-    : m_session(source)
-    , m_initData(WTFMove(initData))
+CDMInstanceOpenCDM::Session::Session(OpenCDMAccessor& source, const String& keySystem, const char* mimeType, Ref<WebCore::SharedBuffer>&& initData, LicenseType licenseType, Ref<WebCore::SharedBuffer>&& customData)
+    : m_initData(WTFMove(initData))
 {
-    // FIXME: This can be a source of overflows and other big errors. This should be fixed in the framework and then fixed here.
-    uint8_t temporaryURL[1024];
-    uint16_t temporaryURLLength = sizeof(temporaryURL);
-
-    std::string message;
-    m_session.GetKeyMessage(message, temporaryURL, temporaryURLLength);
+    auto& callbacksNotifier = OCDMSessionCallbacksNotifier::singleton();
 
-    if (message.empty() || !temporaryURLLength)
+    OpenCDMSession* session = nullptr;
+    opencdm_create_session(&source, keySystem.utf8().data(), openCDMLicenseType(licenseType), mimeType, reinterpret_cast<const uint8_t*>(m_initData->data()), m_initData->size(),
+        !customData->isEmpty() ? reinterpret_cast<const uint8_t*>(customData->data()) : nullptr, customData->size(), callbacksNotifier.callbacks(), &session);
+    if (!session)
         return;
+    m_session = ScopedSession(session, [](OpenCDMSession* session) {
+        opencdm_session_close(session);
+        opencdm_destruct_session(session);
+    });
+    m_id = String::fromUTF8(opencdm_session_id(m_session.get()));
+    callbacksNotifier.addObserver(m_session.get(), this);
+}
 
-    m_isValid = temporaryURLLength;
+CDMInstanceOpenCDM::Session::~Session()
+{
+    OCDMSessionCallbacksNotifier::singleton().removeObserver(this);
+}
 
+void CDMInstanceOpenCDM::Session::challengeGeneratedCallback(String&& message)
+{
     // We could do all operations with String but this way we can save some copies.
     size_t typePosition = message.find(":Type:");
-    String requestType(message.c_str(), typePosition != std::string::npos ? typePosition : 0);
+    String requestType(message.characters8(), typePosition != notFound ? typePosition : 0);
     unsigned offset = 0;
-    if (!requestType.isEmpty() && requestType.length() != message.size())
+    if (!requestType.isEmpty() && requestType.length() != message.length())
         offset = typePosition + 6;
 
-    m_message = SharedBuffer::create(message.c_str() + offset, message.size() - offset);
+    m_message = SharedBuffer::create(message.characters8() + offset, message.length() - offset);
     m_needsIndividualization = requestType.length() == 1 && static_cast<WebCore::MediaKeyMessageType>(requestType.toInt()) == CDMInstance::MessageType::IndividualizationRequest;
+    for (const auto& challengeCallback : m_challengeCallbacks)
+        challengeCallback(this);
+    m_challengeCallbacks.clear();
 }
 
-std::pair<media::OpenCdm::KeyStatus, String> CDMInstanceOpenCDM::Session::update(const uint8_t* data, const unsigned length)
+void CDMInstanceOpenCDM::Session::keyUpdatedCallback(String&& status)
```

Comment:
It'll still have to changed to String at some point and doing it higher is simpky taking adventage of a common place. Otherwise it'll have to be changed here and in challengeGeneratedCallback separately.
Also it doesn't seem to me it can be removed since the callbacks tries to parse something out of it. Mabe I should rename it to response instead?